src/S3IteratorLibsvm.cpp

```diff
@@ -0,0 +1,412 @@
+}
+
+template <class T>
+T S3IteratorLibsvm::readNum(uint64_t& index, std::string& data) {
+  if (!isdigit(data[index])) {
+    throw std::runtime_error("Error in the dataset");
+  }
+
+  uint64_t index_fw = index;
+  // if float or double also accept dots
+  if constexpr (std::is_same<T, double>::value ||
+                std::is_same<T, float>::value) {
+    while (isdigit(data[index_fw]) || data[index_fw] == '.') {
+      index_fw++;
+    }
+  } else {
+    while (isdigit(data[index_fw])) {
+      index_fw++;
+    }
+  }
+
+  char c = data[index_fw];
+  data[index_fw] = 0;
+
+  T result;
+  if constexpr (std::is_same<T, int>::value) {
+    sscanf(&data[index], "%d", &result);
+  } else if constexpr (std::is_same<T, double>::value) {
+    sscanf(&data[index], "%lf", &result);
+  } else if constexpr (std::is_same<T, float>::value) {
+    sscanf(&data[index], "%f", &result);
+  } else if constexpr (std::is_same<T, uint64_t>::value) {
+    sscanf(&data[index], "%lu", &result);
+  } else {
+    throw std::runtime_error("Data type not supported");
+  }
+
+  data[index_fw] = c;  // repair
+  index = index_fw;
+  return result;
+}
+
+bool S3IteratorLibsvm::buildDatasetCsv(
+    const std::string& data,
+    uint64_t index,
+    std::shared_ptr<SparseDataset>& minibatch) {
+  return false;
+}
+
+bool S3IteratorLibsvm::buildDatasetVowpalWabbit(
+    const std::string& data,
+    uint64_t index,
+    std::shared_ptr<SparseDataset>& minibatch) {
+  return false;
+}
+
+/**
+ * Build minibatch from text in libsvm format
+ * We assume index is at a start of a line
+ * index moves forward
+ */
+bool S3IteratorLibsvm::buildDatasetLibsvm(
+    std::string& data,
+    uint64_t& index,
+    std::shared_ptr<SparseDataset>& minibatch) {
+#ifdef DEBUG
+  std::cout << "building dataset libsvm index: " << index << std::endl;
+#endif
+  // libsvm format
+  // <label> <index1>:<value1> <index2>:<value2>
+
+  try {
+    std::vector<std::vector<std::pair<int, FEATURE_TYPE>>> samples;
+    std::vector<FEATURE_TYPE> labels;
+
+    samples.resize(minibatch_rows);
+    labels.resize(minibatch_rows);
+
+    for (uint64_t sample = 0; sample < minibatch_rows; ++sample) {
+      // ignore spaces
+      if (!ignoreSpaces(index, data)) {
+#ifdef DEBUG
+        std::cout << "ignoreSpaces did not find dig index: " << index
+                  << std::endl;
+#endif
+        // did not end up in a digit
+        return false;
+      }
+      int label = readNum<int>(index, data);
+#ifdef DEBUG
+      // std::cout << "read label: " << label << std::endl;
+#endif
+
+      // read pairs
+      while (1) {
+#ifdef DEBUG
+        std::cout << "index: " << index << " reading new pair " << std::endl;
+#endif
+        if (!ignoreSpacesNotNewline(index, data)) {
+          if (data[index] == '\n') {
+#ifdef DEBUG
+            std::cout << "read line index: " << index << std::endl;
+#endif
+            break;  // move to next sample
+          } else if (data[index] == 0) {
```

Comment:
I think this case would never happen. 0 is a digit, so IgnoreSpacesNotNewline would return true if the index was 0.

Reply:
`data[index] == 0`

This is not equal to `data[index] == '0'`.